pallets/pallet-lottery/src/lib.rs

```diff
@@ -0,0 +1,1077 @@
+                if winning_number.saturated_into::<u128>() < number_to_start_rejecting_at {
+                    break;
+                } else {
+                    // sample must be rejected because it can't be safely modulo'd, retry with high u128
+                    winning_number = (as_number >> 128).low_u128();
+                    let number_to_start_rejecting_at: u128 = u128::max_value().saturating_sub(
+                        u128::max_value() % max_winning_balance.saturated_into::<u128>(),
+                    );
+                    debug_assert!(
+                        number_to_start_rejecting_at
+                            <= max_winning_balance.saturated_into::<u128>()
+                    );
+                    if winning_number.saturated_into::<u128>() < number_to_start_rejecting_at {
+                        break;
+                    }
+                    if n + 1 < MAX_NUMBER_OF_RESAMPLES {
+                        // if still not good, we need to re-request randomness with a changed nonce
+                        continue;
+                    } else {
+                        // we loop this up to 3 times (yielding 6 possible values) before giving up, printing a warning and accepting that the result was biased
+                        log::warn!("No unbiased random samples found after {:?} retries. using {:?} which will be subject to modulo bias",(n+1)*2,winning_number.clone());
+                        break;
+                    }
+                }
+            }
+            // no risk of modulo bias here unless we ran out of retries above
+            let winning_balance: BalanceOf<T> = BalanceOf::<T>::try_from(winning_number)
+                .map_err(|_| Error::<T>::ArithmeticOverflow)?
+                % max_winning_balance;
+            log::debug!(
+                "winning_number: {:?}, winning balance: {:?}",
+                winning_number,
+                winning_balance
+            );
+            Ok(winning_balance)
+        }
+        fn select_winner(payout_for_winner: BalanceOf<T>) -> DispatchResult {
+            if Self::total_pot().is_zero() {
+                log::warn!(
+                    "No funds eligible for drawing at block {:?} ",
+                    <frame_system::Pallet<T>>::block_number()
+                );
+                return Ok(());
+            }
+            // Match random number to winner. We select a winning **balance** and then just add up accounts in the order they're stored until the sum of balance exceeds the winning amount
+            // IMPORTANT: This order and active balances must be locked to modification after the random seed is created (relay BABE randomness, 2 epochs ago)
+            let winning_balance = Self::select_winning_balance(Self::total_pot())?;
+            let mut winner: Option<T::AccountId> = None;
+            let mut count: BalanceOf<T> = 0u32.into();
+            for (account, balance) in ActiveBalancePerUser::<T>::iter() {
```

Comment:
maybe we should make this storage item bounded because of this loop ActiveBalancePerUser

Reply:
that would bound the number of users, doesn't seem desirable.
If this needs to be limited then it should be done by the "min_deposit" configurable

Reply:
yeah you're right the min-deposit should help with this. We should maybe add a ci to check that the min-deposit doesn't get something absurdly cheap though

Reply:
@Apokalip could you help with implementing this pls?